sql/src/main/java/io/crate/execution/engine/indexing/GroupRowsByShard.java

```diff
@@ -67,25 +73,41 @@
         this.rowShardResolver = rowShardResolver;
         this.indexNameResolver = indexNameResolver;
         this.expressions = expressions;
+        this.sourceInfoExpressions = sourceInfoExpressions;
         this.itemFactory = itemFactory;
+        this.itemFailureConsumer = itemFailureConsumer;
+        this.possibleSourceUriProcessor = possibleSourceUriProcessor;
         this.autoCreateIndices = autoCreateIndices;
     }
 
     @Override
     public void accept(ShardedRequests<TReq, TItem> shardedRequests, Row row) {
-        rowShardResolver.setNextRow(row);
-        for (int i = 0; i < expressions.size(); i++) {
-            expressions.get(i).setNextRow(row);
+        for (int i = 0; i < sourceInfoExpressions.size(); i++) {
+            sourceInfoExpressions.get(i).setNextRow(row);
         }
-        String id = rowShardResolver.id();
-        TItem item = itemFactory.apply(id);
-        String indexName = indexNameResolver.get();
-        String routing = rowShardResolver.routing();
-        ShardLocation shardLocation = getShardLocation(indexName, id, routing);
-        if (shardLocation == null) {
-            shardedRequests.add(item, indexName, routing);
-        } else {
-            shardedRequests.add(item, shardLocation);
+        if (possibleSourceUriProcessor.apply(shardedRequests)) {
+            // source uri failed processing (reading)
+            return;
+        }
+
+        try {
+            rowShardResolver.setNextRow(row);
+            for (int i = 0; i < expressions.size(); i++) {
+                expressions.get(i).setNextRow(row);
+            }
+
+            String id = rowShardResolver.id();
+            TItem item = itemFactory.apply(id);
+            String indexName = indexNameResolver.get();
+            String routing = rowShardResolver.routing();
+            ShardLocation shardLocation = getShardLocation(indexName, id, routing);
+            if (shardLocation == null) {
+                shardedRequests.add(item, indexName, routing);
+            } else {
+                shardedRequests.add(item, shardLocation);
+            }
+        } catch (Throwable t) {
+            itemFailureConsumer.accept(shardedRequests, t.getMessage());
```

Comment:
Where would this Throwable be thrown?

Reply:
on row collection, on id resolving, etc...